---
title: "Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction"
author: "Cathy Westhues"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
link-citations: true
bibliography: REFERENCES.bib 
vignette: >
  %\VignetteIndexEntry{Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r global_options, include = FALSE}
knitr::opts_chunk$set(comment = "#", collapse = TRUE)
```


# Step 1: Specifying input data and processing parameters
First, we create an object of class \code{METData} with the function [`create_METData()`](https://cathyjubin.github.io/learnMET/reference/create_METData.html).
The user must provide as input data genotypic and phenotypic data, as well as basic information about the field experiments (e.g. longitude, latitude, planting and harvest date), and possibly environmental covariates (if available). These input data are checked and warning messages are given as output if the data are not correctly formatted. \cr
In this example, we use an indica rice dataset from Monteverde et al. (2019), which is implemented in the package as a "toy dataset". 
From this study, a multi-year dataset of rice trials containing phenotypic data (four traits), genotypic and environmental data for a panel of indica genotypes across three years in a single location is available.
(more information on the datasets with `?pheno_indica`,`?geno_indica`,`?map_indica`,`?climate_variables_indica`,`?info_environments_indica`).\cr
In this case, environmental covariates by growth stage are directly available and can be used in predictions. These data should be provided as input in [`create_METData()`] using the argument *env_data*.
Hence, there is no need to retrieve with the package any daily weather data (hence *compute_climatic_ECs* argument set as FALSE).


```{r, message=FALSE, warning=FALSE}
library(learnMET)

data("geno_indica")
data("map_indica")
data("pheno_indica")
data("info_environments_indica")
data("env_data_indica")

METdata_indica <-
  create_METData(
    geno = geno_indica,
    pheno = pheno_indica,
    climate_variables = climate_variables_indica,
    compute_climatic_ECs = F,
    info_environments = info_environments_indica,
    map = map_indica
  )

```

# Step 2: Cross-validated model evaluation


The goal of `predict_trait_MET_cv()` is to assess a given prediction method using a certain type of cross-validation (CV) scenario on the complete training set. The CV schemes covered by the package correspond to those generally evaluated in related literature on MET (@jarquin2014reaction; @jarquin2017increasing;@costa2021nonlinear).\

Here, we will use the CV2: predicting the performance of genotypes in incomplete field trials, meaning that one can use phenotypic information in the training set from other genotypes tested in the same environment, or from the same genotype evalauted in other environments. We also define the percentage of phentoypic observations which should be included in the training set, as well as the number of repetitions. \

The function `predict_trait_MET_cv()` also allows to specify a specific subset of environmental variables from the METData$env_data object to be used in model fitting and predictions via the argument `list_env_predictors`.

**How does `predict_trait_MET_cv()` works?**
When `predict_trait_MET_cv()` is executed, a list of training/test splits is constructed according to the CV scheme chosen by the user. Each training set in each sub-element of this list is processed (e.g. standardization and removal of predictors with null variance.

The function applies a nested CV to obtain an unbiased generalization performance estimate, implying an inner loop CV nested in an outer CV. The inner loop is used for model selection, i.e. hyperparameter tuning with Bayesian optimization, while the outer loop is responsible for evaluation of model performance. Additionnally, there is a possibility for the user to specify a seed to allow reproducibility of analyses. If not provided, a random one is generated and provided in the results file.\
In `predict_trait_MET_cv()`, predictive ability is always calculated within the same environment (locationâ€“year combination), regardless of how the test sets are defined according to the different CV schemes. 


Let's use XGBoost algorithm (@chen2016xgboost) to predict a phenotypic trait!\
**We recommend to use `xgb_reg_1` or `xgb_reg_2` as prediction method, as these methods use principal component as genomic features instead of all SNPs as predictor variables. Hence, the model utilizes less features as input data and is much faster.**

Here we also indicate the number of principal components used as genomic predictor variables, specified via the argument `num_pcs`. All of environmental predictors will be used (`list_env_predictors` is NULL by default.)

```{r eval=FALSE}
rescv0_1 <- predict_trait_MET_cv(
  METData = METdata_indica,
  trait = 'PH',
  prediction_method = 'xgb_reg_1',
  use_selected_markers = F,
  num_pcs = 300,
  lat_lon_included = F,
  year_included = F,
  cv_type = 'cv2',
  nb_folds_cv2 = 4,
  repeats_cv2 = 10,
  include_env_predictors = T,
  save_processing  = T,
  seed = 100,
  path_folder = '~/INDICA/xgb/cv2'
)
```

Note that many methods for processing data based on user-defined parameters and machine learning-based methods are using functions from the tidymodels (https://www.tidymodels.org/) collection of R packages (@tidymodels).

# Step 3: Comes soon!

# Step 4: Comes soon!

# References
