<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Create a METData object with weather data retrieved by nasapower • learnMET</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Create a METData object with weather data retrieved by nasapower">
<meta property="og:description" content="learnMET">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/model_interpretation.html">Interpretation of the fitted ML models using genomic and environmental data</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_stacking_indica.html">Evaluation of the model performance of a stacked model using CV0 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.html">Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_getweatherdata.html">Create a METData object with weather data retrieved by nasapower</a>
    </li>
    <li>
      <a href="../articles/vignette_models_package.html">Overview on the different Machine Learning-based models in learnMET</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_getweatherdata_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Create a METData object with weather data
retrieved by nasapower</h1>
                        <h4 class="author">Cathy Westhues</h4>
            
            <h4 class="date">2022-09-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/cjubin/learnMET/blob/master/vignettes/vignette_getweatherdata.Rmd"><code>vignettes/vignette_getweatherdata.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_getweatherdata.Rmd</code></div>

    </div>

    
    
<div id="specifying-input-data-and-processing-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-input-data-and-processing-parameters" class="anchor"></a>Specifying input data and processing parameters</h2>
<p>In this example, we will show how to create a <code>METData</code>
object when one needs to obtain weather data from external data source.
In the package, weather information from the NASA/POWER is extracted and
aggregated in day-windows which will be thereafter used (in step 2) as
environmental predictor variables.<br>
The user must provide as input data genotypic and phenotypic data, as
well as basic information about the field experiments (e.g. longitude,
latitude, planting and harvest date necessary to retrieve weather data
from the growing season). These input data are checked and warning
messages are given as output if the data are not correctly
formatted.<br>
In this example, we use data from the <a href="https://www.genomes2fields.org/">Genomes to Fields dataset</a>
(<span class="citation">AlKhalifah et al. (<a href="#ref-alkhalifah2018maize" role="doc-biblioref">2018</a>)</span>,
<span class="citation">McFarland et al. (<a href="#ref-mcfarland2020maize" role="doc-biblioref">2020</a>)</span>),
which is implemented in the package as a toy dataset. For this dataset,
soil variables are available <strong>and we will retrieve weather data
using the nasapower package</strong> (<span class="citation">Sparks (<a href="#ref-sparks2018nasapower" role="doc-biblioref">2018</a>)</span>)
for the 22 environments included in the dataset. (more information on
the datasets with
<code><a href="../reference/pheno_G2F.html">?pheno_G2F</a></code>,<code><a href="../reference/geno_G2F.html">?geno_G2F</a></code>,<code><a href="../reference/map_G2F.html">?map_G2F</a></code>,<code>?climate_variables_G2F</code>,<code><a href="../reference/info_environments_G2F.html">?info_environments_G2F</a></code>,
<code><a href="../reference/soil_G2F.html">?soil_G2F</a></code>).<br>
In this case, soil covariates by environment are available and can be
used in predictions. These data should be provided as input in
<code><a href="../reference/create_METData.html">create_METData()</a></code> using the argument
<em>soil_variables</em>. To get weather data from NASA/POWER, one must
set the argument <em>compute_climatic_ECs</em> to TRUE.<br>
Do not forget to indicate where the daily weather data and the plots of
clustering analyses should be saved using the argument
<em>path_to_save</em>.<br>
The daily weather data will be saved in a subdirectory of this directory
that is automatically created and called
<strong>weather_data</strong>.<br>
The method used to build environmental covariates based on daily weather
data is the default one: “fixed_nb_windows_across_env”, which means that
in each environment, the growing season is split into a number of
windows equal to nb_windows_intervals (by defaultm this value is equal
to 10). To get more information on the different methods implemented in
the package to build climatic environmental variables based ond daily
weather information, please refer to the documentation of the function
<code><a href="../reference/get_ECs.html">get_ECs()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cjubin/learnMET">learnMET</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">geno_G2F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">pheno_G2F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">map_G2F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">info_environments_G2F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">soil_G2F</span><span class="op">)</span>
<span class="va">METdata_G2F</span> <span class="op">&lt;-</span>
  <span class="fu">learnMET</span><span class="fu">::</span><span class="fu"><a href="../reference/create_METData.html">create_METData</a></span><span class="op">(</span>
    geno <span class="op">=</span> <span class="va">geno_G2F</span>,
    pheno <span class="op">=</span> <span class="va">pheno_G2F</span>,
    map <span class="op">=</span> <span class="va">map_G2F</span>,
    climate_variables <span class="op">=</span> <span class="cn">NULL</span>,
    compute_climatic_ECs <span class="op">=</span> <span class="cn">TRUE</span>,
    info_environments <span class="op">=</span> <span class="va">info_environments_G2F</span>,
    soil_variables <span class="op">=</span> <span class="va">soil_G2F</span>,
    path_to_save <span class="op">=</span> <span class="st">'~/learnMET_analyses/G2F'</span>
  <span class="op">)</span>
<span class="co">## No climate covariates provided by the user.</span>
<span class="co">## Computation of environmental covariates starts.</span>
<span class="co">## Daily weather tables have been downloaded from NASA POWER for the required environments in a previous run, and are matching the environments ID/planting and harvest dates used in this analysis.</span>
<span class="co">##  These data will be used. </span>
<span class="co">## Daily weather tables downloaded from NASA POWER for the required environments!</span>
<span class="co">## Computation of environmental covariates is done.</span>
<span class="co">## Clustering of env. data starts.</span>
<span class="co">## Clustering of env. data done.</span>
<span class="co">## Soil and climate data will be included in the final METData object.</span></code></pre></div>
<p>The function <code><a href="../reference/print.summary.METData.html">print.summary.METData()</a></code> gives an overview
of the created MET data object.<br></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">METdata_G2F</span><span class="op">)</span>
<span class="co">## object of class 'METData' </span>
<span class="co">## --------------------------</span>
<span class="co">## General information about the MET data </span>
<span class="co">## </span>
<span class="co">## No. of unique environments represented in the data: 22 </span>
<span class="co">##        unique years represented in the data: 4 </span>
<span class="co">##        unique locations represented in the data: 6 </span>
<span class="co">## </span>
<span class="co">## Distribution of phenotypic observations according to year and location: </span>
<span class="co">##       </span>
<span class="co">##        Aurora CollegeStation Columbia Georgetown Waterloo WestMadison</span>
<span class="co">##   2014    212            236      244        224      181         200</span>
<span class="co">##   2015    347            290      411        407      345           0</span>
<span class="co">##   2016    228            197      328        227      207         303</span>
<span class="co">##   2017    187              0      217        163      171         266</span>
<span class="co">## No. of unique genotypes which are phenotyped</span>
<span class="co">##  1913 </span>
<span class="co">## --------------------------</span>
<span class="co">## Climate variables </span>
<span class="co">## Weather data extracted from NASAPOWER for main weather variables?</span>
<span class="co">## YES</span>
<span class="co">## No. of climate variables available: 130 </span>
<span class="co">## --------------------------</span>
<span class="co">## Soil variables </span>
<span class="co">## No. of soil variables available: 4 </span>
<span class="co">## --------------------------</span>
<span class="co">## Phenotypic data </span>
<span class="co">##   No. of traits:  3 </span>
<span class="co">## </span>
<span class="co">##      pltht         yld_bu_ac           earht      </span>
<span class="co">##  Min.   :117.2   Min.   :  2.814   Min.   : 48.0  </span>
<span class="co">##  1st Qu.:206.0   1st Qu.:114.641   1st Qu.:100.5  </span>
<span class="co">##  Median :232.5   Median :152.345   Median :115.6  </span>
<span class="co">##  Mean   :227.2   Mean   :147.758   Mean   :114.0  </span>
<span class="co">##  3rd Qu.:251.0   3rd Qu.:181.874   3rd Qu.:128.5  </span>
<span class="co">##  Max.   :351.0   Max.   :344.631   Max.   :195.0  </span>
<span class="co">##  NA's   :43                        NA's   :1072   </span>
<span class="co">## --------------------------</span>
<span class="co">## Genotypic data </span>
<span class="co">##   No. of markers : 106414 </span>
<span class="co">## --------------------------</span>
<span class="co">## Map data</span>
<span class="co">##   No. of chromosomes     10 </span>
<span class="co">## </span>
<span class="co">##   markers per chromosome </span>
<span class="co">##  </span>
<span class="co">##     1     2     3     4     5     6     7     8     9    10 </span>
<span class="co">## 16815 12533 11852  9507 13658  8151  8414  8936  8273  8275</span></code></pre></div>
</div>
<div id="looking-at-the-results-from-the-clustering-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#looking-at-the-results-from-the-clustering-analyses" class="anchor"></a>Looking at the results from the clustering analyses</h2>
<p>The <code><a href="../reference/create_METData.html">create_METData()</a></code> function integrates clustering
analyses with the kmeans algorithm (<span class="citation">Hartigan and
Wong (<a href="#ref-hartigan1979algorithm" role="doc-biblioref">1979</a>)</span>) on environmental variables.<br>
In the directory corresponding to the <em>path_to_save</em> argument, a
subdirectory called <strong>clustering_analysis</strong> is created,
which contains 3 additional subdirectories:
<strong>only_weather</strong>, which contains clustering plots based on
solely weather-based variables; <strong>only_soil</strong>, which
contains clustering plots based on solely soil variables; and
<strong>all_env_data</strong>, which contains clustering plots based on
all environmental variables (soil + weather-based).<br>
All clustering analyses are performed with different K (number of
clusters), and corresponding plots are generated for a range of k
values. In addition, plots to visualize the total within-cluster sums of
squares (elbow method) and the average Silhouette score (<span class="citation">Hartigan and Wong (<a href="#ref-hartigan1979algorithm" role="doc-biblioref">1979</a>)</span>) - metric obtained with the
cluster package (<span class="citation">Maechler et al. (<a href="#ref-clusterpackage" role="doc-biblioref">2021</a>)</span>)) for a
range of k values are also saved in the adequate folders. These latter
plots can be useful to assess the best number of clusters to
consider.<br>
NOte that for the clustering analyses are run with
<code>nstart = 25</code> so that the k-means algorithm create multiple
initial configurations and outputs the best one.<br>
For instance, with k = 4,below is the plot of how environments included
in this MET dataset cluster, with k = 4:
<img src="climate_variables_only_clusters_environments_4.png" width="2304"><br>
Below is the plot obtained for the elbow method with different k values:
one should decide on the number of clusters so that, when adding another
cluster, it does not improve much the total intra-cluster variation
(that we want to minimize). In this case, 4 seems to be not so
bad!<br><img src="plot_elbow_method.png" width="307"><br>
The Silhouette score is also used to validate results from clustering.
It gives a measure of how well observations are clustered with other
observations within a cluster. It is calculated for each observation of
the dataset, and the final metric corresponds to the average over all
observations for a given k. A large average Silhouette score reflects
that observations are well clustered with this k number of
clusters.<br>
Here, 5 seems to be a better choice based on the Silhouette score.
<img src="plot_sil_score.png" width="307"></p>
</div>
<div id="computing-environmental-covariates-different-methods-possible" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-environmental-covariates-different-methods-possible" class="anchor"></a>Computing environmental covariates: different methods possible</h2>
<p>In the first chunk of code presented above, daily weather data have
been saved in a folder called <strong>weather_data</strong>. We can load
these daily weather tables from NASA/POWER, and then implement different
methods to build day-windows (using the argument
<em>method_ECs_intervals</em>) which will represent climate variables
for the ML-based prediction models. To re-use the daily weather data
obtained via the NASA/POWER query, just specify these with the argument
<em>raw_weather_data</em> (see example below). These data will be
analyzed as<br>
For instance with “fixed_length_time_windows_across_env” as below, with
day-windows computed in each environment with a width of 8 days:<br><strong>Note: be careful to use a different path so that the clustering
analyses results obtained with the different environmental datasets do
not overlap!</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">daily_weather_tables_g2f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"daily_weather_tables_nasapower.RDS"</span><span class="op">)</span>

<span class="va">METdata_G2F_2</span> <span class="op">&lt;-</span>
  <span class="fu">learnMET</span><span class="fu">::</span><span class="fu"><a href="../reference/create_METData.html">create_METData</a></span><span class="op">(</span>
    geno <span class="op">=</span> <span class="va">geno_G2F</span>,
    pheno <span class="op">=</span> <span class="va">pheno_G2F</span>,
    map <span class="op">=</span> <span class="va">map_G2F</span>,
    climate_variables <span class="op">=</span> <span class="cn">NULL</span>,
    compute_climatic_ECs <span class="op">=</span> <span class="cn">TRUE</span>,
    raw_weather_data <span class="op">=</span> <span class="va">daily_weather_tables_g2f</span>,
    method_ECs_intervals <span class="op">=</span> <span class="st">"fixed_length_time_windows_across_env"</span>,
    duration_time_window_days <span class="op">=</span> <span class="fl">8</span>,
    info_environments <span class="op">=</span> <span class="va">info_environments_G2F</span>,
    soil_variables <span class="op">=</span> <span class="va">soil_G2F</span>,
    path_to_save <span class="op">=</span> <span class="st">'~/learnMET_analyses/G2F/method_2'</span>
  <span class="op">)</span>
<span class="co">## No climate covariates provided by the user.</span>
<span class="co">## Computation of environmental covariates starts.</span>
<span class="co">## Raw weather data are provided by the user and will be used to build environmental covariates.</span>
<span class="co">## QC on daily weather data starts...</span>
<span class="co">## QC on daily weather data is done!</span>
<span class="co">## A file with flagged values has been saved in the subfolder weather_data.</span>
<span class="co">## Computation of environmental covariates is done.</span>
<span class="co">## Clustering of env. data starts.</span>
<span class="co">## Clustering of env. data done.</span>
<span class="co">## Soil and climate data will be included in the final METData object.</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">METdata_G2F_2</span><span class="op">)</span>
<span class="co">## object of class 'METData' </span>
<span class="co">## --------------------------</span>
<span class="co">## General information about the MET data </span>
<span class="co">## </span>
<span class="co">## No. of unique environments represented in the data: 22 </span>
<span class="co">##        unique years represented in the data: 4 </span>
<span class="co">##        unique locations represented in the data: 6 </span>
<span class="co">## </span>
<span class="co">## Distribution of phenotypic observations according to year and location: </span>
<span class="co">##       </span>
<span class="co">##        Aurora CollegeStation Columbia Georgetown Waterloo WestMadison</span>
<span class="co">##   2014    212            236      244        224      181         200</span>
<span class="co">##   2015    347            290      411        407      345           0</span>
<span class="co">##   2016    228            197      328        227      207         303</span>
<span class="co">##   2017    187              0      217        163      171         266</span>
<span class="co">## No. of unique genotypes which are phenotyped</span>
<span class="co">##  1913 </span>
<span class="co">## --------------------------</span>
<span class="co">## Climate variables </span>
<span class="co">## Weather data extracted from NASAPOWER for main weather variables?</span>
<span class="co">## NO</span>
<span class="co">## No. of climate variables available: 208 </span>
<span class="co">## --------------------------</span>
<span class="co">## Soil variables </span>
<span class="co">## No. of soil variables available: 4 </span>
<span class="co">## --------------------------</span>
<span class="co">## Phenotypic data </span>
<span class="co">##   No. of traits:  3 </span>
<span class="co">## </span>
<span class="co">##      pltht         yld_bu_ac           earht      </span>
<span class="co">##  Min.   :117.2   Min.   :  2.814   Min.   : 48.0  </span>
<span class="co">##  1st Qu.:206.0   1st Qu.:114.641   1st Qu.:100.5  </span>
<span class="co">##  Median :232.5   Median :152.345   Median :115.6  </span>
<span class="co">##  Mean   :227.2   Mean   :147.758   Mean   :114.0  </span>
<span class="co">##  3rd Qu.:251.0   3rd Qu.:181.874   3rd Qu.:128.5  </span>
<span class="co">##  Max.   :351.0   Max.   :344.631   Max.   :195.0  </span>
<span class="co">##  NA's   :43                        NA's   :1072   </span>
<span class="co">## --------------------------</span>
<span class="co">## Genotypic data </span>
<span class="co">##   No. of markers : 106414 </span>
<span class="co">## --------------------------</span>
<span class="co">## Map data</span>
<span class="co">##   No. of chromosomes     10 </span>
<span class="co">## </span>
<span class="co">##   markers per chromosome </span>
<span class="co">##  </span>
<span class="co">##     1     2     3     4     5     6     7     8     9    10 </span>
<span class="co">## 16815 12533 11852  9507 13658  8151  8414  8936  8273  8275</span></code></pre></div>
<p>If one has information about the phenological stages in each
environment (either by observations done from the field, or from results
from a crop growth model, for instance Sirius (<span class="citation">Jamieson et al. (<a href="#ref-jamieson1998sirius" role="doc-biblioref">1998</a>)</span>) or APSIM (<span class="citation">Keating et al. (<a href="#ref-keating2003overview" role="doc-biblioref">2003</a>)</span>, <span class="citation">Holzworth
et al. (<a href="#ref-holzworth2014apsim" role="doc-biblioref">2014</a>)</span>), one can provide a table which
indicates at which day after planting the crop has reached a new
developmental stage. This table should be provided as input via the
argument <em>intervals_growth_manual</em>, and the method for computing
climate variables should now be chosen as
“user_defined_intervals”.<br>
The names of each developmental stage can be indicated in the column
names, which can be useful after analyses to identify which wether-based
covariate had more weight in the machine learning model used. An example
of a table is given as example data in the package (see below).
<strong>The example provided is given to illustrate how the user should
format the table, but was not created to reflect the true growth stage
of the crop in these environments!</strong><br></p>
<p>In this case, each growth period is estimated for <strong>all
genotypes tested in this environment</strong>: the package does not
incorporate at the moment specificity at the genotype level to take into
account genotypic variation regarding plant development and phenology.
For more information on this topic, please have a look at the following
publications: <span class="citation">Heslot et al. (<a href="#ref-heslot2014integrating" role="doc-biblioref">2014</a>)</span>,
<span class="citation">Technow et al. (<a href="#ref-technow2015integrating" role="doc-biblioref">2015</a>)</span>, <span class="citation">Rincent et
al. (<a href="#ref-rincent2019using" role="doc-biblioref">2019</a>)</span>.<br></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"intervals_growth_manual_G2F"</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">intervals_growth_manual_G2F</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">year</th>
<th align="left">location</th>
<th align="right">P</th>
<th align="right">VE</th>
<th align="right">V5</th>
<th align="right">V12</th>
<th align="right">R1</th>
<th align="right">R3</th>
<th align="right">R6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2014</td>
<td align="left">Aurora</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
<tr class="even">
<td align="right">2014</td>
<td align="left">CollegeStation</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
<tr class="odd">
<td align="right">2014</td>
<td align="left">Columbia</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
<tr class="even">
<td align="right">2014</td>
<td align="left">Georgetown</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
<tr class="odd">
<td align="right">2014</td>
<td align="left">Waterloo</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
<tr class="even">
<td align="right">2014</td>
<td align="left">WestMadison</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">21</td>
<td align="right">44</td>
<td align="right">63</td>
<td align="right">91</td>
<td align="right">119</td>
</tr>
</tbody>
</table>
<p>One can also modify the base temperature (using the argument
<em>base_temperature</em> - see example below) which is used to estimate
accumulated GDD in each environment. The value of the base temperature
depends on the crop under study and can be found in the literature (see
<a href="https://en.wikipedia.org/wiki/Growing_degree-day#Baselines" class="uri">https://en.wikipedia.org/wiki/Growing_degree-day#Baselines</a>
to get the base temperature for widely used crops).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">METdata_G2F_2</span> <span class="op">&lt;-</span>
  <span class="fu">learnMET</span><span class="fu">::</span><span class="fu"><a href="../reference/create_METData.html">create_METData</a></span><span class="op">(</span>
    geno <span class="op">=</span> <span class="va">geno_G2F</span>,
    pheno <span class="op">=</span> <span class="va">pheno_G2F</span>,
    map <span class="op">=</span> <span class="va">map_G2F</span>,
    climate_variables <span class="op">=</span> <span class="cn">NULL</span>,
    compute_climatic_ECs <span class="op">=</span> <span class="cn">TRUE</span>,
    raw_weather_data <span class="op">=</span> <span class="va">daily_weather_tables_g2f</span>,
    method_ECs_intervals <span class="op">=</span> <span class="st">"user_defined_intervals"</span>,
    intervals_growth_manual <span class="op">=</span> <span class="va">intervals_growth_manual_G2F</span>,
    base_temperature <span class="op">=</span> <span class="fl">12</span>,
    info_environments <span class="op">=</span> <span class="va">info_environments_G2F</span>,
    soil_variables <span class="op">=</span> <span class="va">soil_G2F</span>,
    path_to_save <span class="op">=</span> <span class="st">'~/learnMET_analyses/G2F/method_growth_stages'</span>
  <span class="op">)</span>
<span class="co">## No climate covariates provided by the user.</span>
<span class="co">## Computation of environmental covariates starts.</span>
<span class="co">## Raw weather data are provided by the user and will be used to build environmental covariates.</span>
<span class="co">## QC on daily weather data starts...</span>
<span class="co">## QC on daily weather data is done!</span>
<span class="co">## A file with flagged values has been saved in the subfolder weather_data.</span>
<span class="co">## Computation of environmental covariates is done.</span>
<span class="co">## Clustering of env. data starts.</span>
<span class="co">## Clustering of env. data done.</span>
<span class="co">## Soil and climate data will be included in the final METData object.</span></code></pre></div>
</div>
<div id="getting-soil-data-from-public-database-soilgrids" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-soil-data-from-public-database-soilgrids" class="anchor"></a>Getting soil data from public database (SoilGrids)</h2>
<p>If no soil data is provided by the user, it is possible to retrieve
publicly available data from the SoilGrids database (<span class="citation">Poggio et al. (<a href="#ref-poggio2021soilgrids" role="doc-biblioref">2021</a>)</span>). In this case, one needs to set
to <code>TRUE</code> the argument <code>get_public_soil_data</code>. In
this example, we use the multi-environment rice dataset deployed as a
toy dataset in learnMET (<span class="citation">Monteverde et al. (<a href="#ref-monteverde2019integrating" role="doc-biblioref">2019</a>)</span>):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"geno_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"map_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"pheno_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"info_environments_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"climate_variables_indica"</span><span class="op">)</span>

<span class="va">METdata_indica</span> <span class="op">&lt;-</span>
    <span class="fu">learnMET</span><span class="fu">::</span><span class="fu"><a href="../reference/create_METData.html">create_METData</a></span><span class="op">(</span>
        geno <span class="op">=</span> <span class="va">geno_indica</span>,
        pheno <span class="op">=</span> <span class="va">pheno_indica</span>,
        climate_variables <span class="op">=</span> <span class="va">climate_variables_indica</span>,
        compute_climatic_ECs <span class="op">=</span> <span class="cn">F</span>,
        info_environments <span class="op">=</span> <span class="va">info_environments_indica</span>,
        map <span class="op">=</span> <span class="va">map_indica</span>,
        get_public_soil_data <span class="op">=</span> <span class="cn">T</span>,
        path_to_save <span class="op">=</span> <span class="st">"~/learnMET_analyses/indica_rice_data"</span>
    <span class="op">)</span>
<span class="co">## No soil covariates provided by the user.</span>
<span class="co">## The package will retrieve soil data from the SoilGrids (ISRIC) Database.</span>
<span class="co">## Clustering of env. data starts.</span>
<span class="co">## Only one location used so no variance in the soil predictors. No clustering analyses possible.</span>
<span class="co">## Clustering of env. data done.</span>
<span class="co">## Soil and climate data will be included in the final METData object.</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">METdata_indica</span><span class="op">$</span><span class="va">env_data</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">METdata_indica</span><span class="op">$</span><span class="va">list_soil_predictors</span><span class="op">)</span><span class="op">)</span>
<span class="co">##  [1] "silt_0-5cm"        "silt_5-15cm"       "silt_15-30cm"     </span>
<span class="co">##  [4] "silt_30-60cm"      "silt_60-100cm"     "clay_0-5cm"       </span>
<span class="co">##  [7] "clay_5-15cm"       "clay_15-30cm"      "clay_30-60cm"     </span>
<span class="co">## [10] "clay_60-100cm"     "sand_0-5cm"        "sand_5-15cm"      </span>
<span class="co">## [13] "sand_15-30cm"      "sand_30-60cm"      "sand_60-100cm"    </span>
<span class="co">## [16] "bdod_0-5cm"        "bdod_5-15cm"       "bdod_15-30cm"     </span>
<span class="co">## [19] "bdod_30-60cm"      "bdod_60-100cm"     "cec_0-5cm"        </span>
<span class="co">## [22] "cec_5-15cm"        "cec_15-30cm"       "cec_30-60cm"      </span>
<span class="co">## [25] "cec_60-100cm"      "nitrogen_0-5cm"    "nitrogen_5-15cm"  </span>
<span class="co">## [28] "nitrogen_15-30cm"  "nitrogen_30-60cm"  "nitrogen_60-100cm"</span>
<span class="co">## [31] "phh2o_0-5cm"       "phh2o_5-15cm"      "phh2o_15-30cm"    </span>
<span class="co">## [34] "phh2o_30-60cm"     "phh2o_60-100cm"    "soc_0-5cm"        </span>
<span class="co">## [37] "soc_5-15cm"        "soc_15-30cm"       "soc_30-60cm"      </span>
<span class="co">## [40] "soc_60-100cm"</span></code></pre></div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-alkhalifah2018maize" class="csl-entry">
AlKhalifah, Naser, Darwin A Campbell, Celeste M Falcon, Jack M Gardiner,
Nathan D Miller, Maria Cinta Romay, Ramona Walls, et al. 2018.
<span>“Maize Genomes to Fields: 2014 and 2015 Field Season Genotype,
Phenotype, Environment, and Inbred Ear Image Datasets.”</span> <em>BMC
Research Notes</em> 11 (1): 1–5.
</div>
<div id="ref-hartigan1979algorithm" class="csl-entry">
Hartigan, John A, and Manchek A Wong. 1979. <span>“Algorithm AS 136: A
k-Means Clustering Algorithm.”</span> <em>Journal of the Royal
Statistical Society. Series c (Applied Statistics)</em> 28 (1): 100–108.
</div>
<div id="ref-heslot2014integrating" class="csl-entry">
Heslot, Nicolas, Deniz Akdemir, Mark E Sorrells, and Jean-Luc Jannink.
2014. <span>“Integrating Environmental Covariates and Crop Modeling into
the Genomic Selection Framework to Predict Genotype by Environment
Interactions.”</span> <em>Theoretical and Applied Genetics</em> 127 (2):
463–80.
</div>
<div id="ref-holzworth2014apsim" class="csl-entry">
Holzworth, Dean P, Neil I Huth, Peter G deVoil, Eric J Zurcher, Neville
I Herrmann, Greg McLean, Karine Chenu, et al. 2014.
<span>“APSIM–Evolution Towards a New Generation of Agricultural Systems
Simulation.”</span> <em>Environmental Modelling &amp; Software</em> 62:
327–50.
</div>
<div id="ref-jamieson1998sirius" class="csl-entry">
Jamieson, PD, MA Semenov, IR Brooking, and GS Francis. 1998.
<span>“Sirius: A Mechanistic Model of Wheat Response to Environmental
Variation.”</span> <em>European Journal of Agronomy</em> 8 (3-4):
161–79.
</div>
<div id="ref-keating2003overview" class="csl-entry">
Keating, Brian A, Peter S Carberry, Graeme L Hammer, Mervyn E Probert,
Michael J Robertson, D Holzworth, Neil I Huth, et al. 2003. <span>“An
Overview of APSIM, a Model Designed for Farming Systems
Simulation.”</span> <em>European Journal of Agronomy</em> 18 (3-4):
267–88.
</div>
<div id="ref-clusterpackage" class="csl-entry">
Maechler, Martin, Peter Rousseeuw, Anja Struyf, Mia Hubert, and Kurt
Hornik. 2021. <em>Cluster: Cluster Analysis Basics and Extensions</em>.
<a href="https://CRAN.R-project.org/package=cluster">https://CRAN.R-project.org/package=cluster</a>.
</div>
<div id="ref-mcfarland2020maize" class="csl-entry">
McFarland, Bridget A, Naser AlKhalifah, Martin Bohn, Jessica Bubert,
Edward S Buckler, Ignacio Ciampitti, Jode Edwards, et al. 2020.
<span>“Maize Genomes to Fields (G2f): 2014–2017 Field Seasons: Genotype,
Phenotype, Climatic, Soil, and Inbred Ear Image Datasets.”</span>
<em>BMC Research Notes</em> 13 (1): 1–6.
</div>
<div id="ref-monteverde2019integrating" class="csl-entry">
Monteverde, Eliana, Lucı́a Gutierrez, Pedro Blanco, Fernando Pérez de
Vida, Juan E Rosas, Victoria Bonnecarrère, Gastón Quero, and Susan
McCouch. 2019. <span>“Integrating Molecular Markers and Environmental
Covariates to Interpret Genotype by Environment Interaction in Rice
(Oryza Sativa l.) Grown in Subtropical Areas.”</span> <em>G3: Genes,
Genomes, Genetics</em> 9 (5): 1519–31.
</div>
<div id="ref-poggio2021soilgrids" class="csl-entry">
Poggio, Laura, Luis M De Sousa, Niels H Batjes, Gerard Heuvelink, Bas
Kempen, Eloi Ribeiro, and David Rossiter. 2021. <span>“SoilGrids 2.0:
Producing Soil Information for the Globe with Quantified Spatial
Uncertainty.”</span> <em>Soil</em> 7 (1): 217–40.
</div>
<div id="ref-rincent2019using" class="csl-entry">
Rincent, Renaud, Marcos Malosetti, Behnam Ababaei, Gaetan Touzy, Agathe
Mini, M Bogard, Pierre Martre, Jacques Le Gouis, and F van Eeuwijk.
2019. <span>“Using Crop Growth Model Stress Covariates and AMMI
Decomposition to Better Predict Genotype-by-Environment
Interactions.”</span> <em>Theoretical and Applied Genetics</em> 132
(12): 3399–3411.
</div>
<div id="ref-sparks2018nasapower" class="csl-entry">
Sparks, Adam H. 2018. <span>“Nasapower: A NASA POWER Global Meteorology,
Surface Solar Energy and Climatology Data Client for r.”</span>
<em>Journal of Open Source Software, 3 (30), 1035</em>.
</div>
<div id="ref-technow2015integrating" class="csl-entry">
Technow, Frank, Carlos D Messina, L Radu Totir, and Mark Cooper. 2015.
<span>“Integrating Crop Growth Models with Whole Genome Prediction
Through Approximate Bayesian Computation.”</span> <em>PloS One</em> 10
(6): e0130855.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
