<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction • learnMET</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction">
<meta property="og:description" content="learnMET">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/model_interpretation.html">Interpretation of the fitted ML models using genomic and environmental data</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_stacking_indica.html">Evaluation of the model performance of a stacked model using CV0 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.html">Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_getweatherdata.html">Create a METData object with weather data retrieved by nasapower</a>
    </li>
    <li>
      <a href="../articles/vignette_models_package.html">Overview on the different Machine Learning-based models in learnMET</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_cv_xgb_indica_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Evaluation of the performance of a XGBoost model
using CV2 on a rice dataset for phenotypic prediction</h1>
                        <h4 class="author">Cathy Westhues</h4>
            
            <h4 class="date">2022-09-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/cjubin/learnMET/blob/master/vignettes/vignette_cv_xgb_indica.Rmd"><code>vignettes/vignette_cv_xgb_indica.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_cv_xgb_indica.Rmd</code></div>

    </div>

    
    
<div id="step-1-specifying-input-data-and-processing-parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#step-1-specifying-input-data-and-processing-parameters" class="anchor"></a>Step 1: Specifying input data and processing parameters</h1>
<p>First, we create an object of class with the function <a href="https://cathyjubin.github.io/learnMET/reference/create_METData.html"><code>create_METData()</code></a>.
The user must provide as input data genotypic and phenotypic data, as
well as basic information about the field experiments (e.g. longitude,
latitude data at least), and possibly environmental covariates (if
available). These input data are checked and warning messages are given
as output if the data are not correctly formatted.<br>
In this example, we use an indica rice dataset from Monteverde et
al. (2019), which is implemented in the package as a “toy dataset”. From
this study, a multi-year dataset of rice trials containing phenotypic
data (four traits), genotypic and environmental data for a panel of
indica genotypes across three years in a single location is available.
(more information on the datasets with
<code><a href="../reference/pheno_indica.html">?pheno_indica</a></code>,<code><a href="../reference/geno_indica.html">?geno_indica</a></code>,<code><a href="../reference/map_indica.html">?map_indica</a></code>,<code><a href="../reference/climate_variables_indica.html">?climate_variables_indica</a></code>,<code><a href="../reference/info_environments_indica.html">?info_environments_indica</a></code>).<br>
In this case, environmental covariates by growth stage are directly
available and can be used in predictions. These data should be provided
as input in [<code><a href="../reference/create_METData.html">create_METData()</a></code>] using the argument
<em>climate_variables</em>. Hence, there is no need to retrieve with the
package any daily weather data (hence <em>compute_climatic_ECs</em>
argument set as FALSE).<br>
Do not forget to indicate where the plots of clustering analyses should
be saved using the argument <em>path_to_save</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cjubin/learnMET">learnMET</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"geno_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"map_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"pheno_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"info_environments_indica"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"climate_variables_indica"</span><span class="op">)</span>

<span class="va">METdata_indica</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/create_METData.html">create_METData</a></span><span class="op">(</span>
    geno <span class="op">=</span> <span class="va">geno_indica</span>,
    pheno <span class="op">=</span> <span class="va">pheno_indica</span>,
    climate_variables <span class="op">=</span> <span class="va">climate_variables_indica</span>,
    compute_climatic_ECs <span class="op">=</span> <span class="cn">F</span>,
    info_environments <span class="op">=</span> <span class="va">info_environments_indica</span>,
    map <span class="op">=</span> <span class="va">map_indica</span>,
    path_to_save <span class="op">=</span> <span class="st">'~/learnMET_analyses/indica/xgb'</span>
  <span class="op">)</span>
<span class="co"># No soil covariates provided by the user.</span></code></pre></div>
</div>
<div id="step-2-cross-validated-model-evaluation" class="section level1">
<h1 class="hasAnchor">
<a href="#step-2-cross-validated-model-evaluation" class="anchor"></a>Step 2: Cross-validated model evaluation</h1>
<div id="specifying-processing-parameters-and-the-type-of-prediction-model" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-processing-parameters-and-the-type-of-prediction-model" class="anchor"></a>Specifying processing parameters and the type of prediction
model</h2>
<p>The goal of <code><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code> is to assess a given
prediction method using a certain type of cross-validation (CV) scenario
on the complete training set. The CV schemes covered by the package
correspond to those generally evaluated in related literature on MET
(<span class="citation">Jarquı́n et al. (<a href="#ref-jarquin2014reaction" role="doc-biblioref">2014</a>)</span>;
<span class="citation">Jarquı́n et al. (<a href="#ref-jarquin2017increasing" role="doc-biblioref">2017</a>)</span>;<span class="citation">Costa-Neto,
Fritsche-Neto, and Crossa (<a href="#ref-costa2021nonlinear" role="doc-biblioref">2021</a>)</span>).<br></p>
<p>Here, we will use the CV2: predicting the performance of genotypes in
incomplete field trials, meaning that one can use phenotypic information
in the training set from other genotypes tested in the same environment,
or from the same genotype evalauted in other environments. We also
define the percentage of phentoypic observations which should be
included in the training set, as well as the number of
repetitions.<br></p>
<p>The function <code><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code> also allows to
specify a specific subset of environmental variables from the
METData$env_data object to be used in model fitting and predictions via
the argument <code>list_env_predictors</code>.</p>
<p><strong>How does <code><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code> works?</strong>
When <code><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code> is executed, a list of
training/test splits is constructed according to the CV scheme chosen by
the user. Each training set in each sub-element of this list is
processed (e.g. standardization and removal of predictors with null
variance.</p>
<p>The function applies a nested CV to obtain an unbiased generalization
performance estimate, implying an inner loop CV nested in an outer CV.
The inner loop is used for model selection, i.e. hyperparameter tuning
with Bayesian optimization, while the outer loop is responsible for
evaluation of model performance. Additionnally, there is a possibility
for the user to specify a seed to allow reproducibility of analyses. If
not provided, a random one is generated and provided in the results
file.<br>
In <code><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code>, predictive ability is always
calculated within the same environment (location–year combination),
regardless of how the test sets are defined according to the different
CV schemes.</p>
<p>Let’s use XGBoost algorithm (<span class="citation">Chen and Guestrin
(<a href="#ref-chen2016xgboost" role="doc-biblioref">2016</a>)</span>)
to predict a phenotypic trait!<br><strong>We recommend to use <code>xgb_reg_1</code> or
<code>xgb_reg_2</code> as prediction method, as these methods use
principal component as genomic features instead of all SNPs as predictor
variables. Hence, the model utilizes less features as input data and is
much faster.</strong></p>
<p>Here we also indicate the number of principal components used as
genomic predictor variables, specified via the argument
<code>num_pcs</code>. All of environmental predictors will be used
(<code>list_env_predictors</code> is NULL by default.)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rescv0_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_trait_MET_cv.html">predict_trait_MET_cv</a></span><span class="op">(</span>
  METData <span class="op">=</span> <span class="va">METdata_indica</span>,
  trait <span class="op">=</span> <span class="st">'GC'</span>,
  prediction_method <span class="op">=</span> <span class="st">'xgb_reg_1'</span>,
  use_selected_markers <span class="op">=</span> <span class="cn">F</span>,
  num_pcs <span class="op">=</span> <span class="fl">80</span>,
  lat_lon_included <span class="op">=</span> <span class="cn">F</span>,
  year_included <span class="op">=</span> <span class="cn">F</span>,
  cv_type <span class="op">=</span> <span class="st">'cv2'</span>,
  nb_folds_cv2 <span class="op">=</span> <span class="fl">4</span>,
  repeats_cv2 <span class="op">=</span> <span class="fl">10</span>,
  include_env_predictors <span class="op">=</span> <span class="cn">T</span>,
  save_processing  <span class="op">=</span> <span class="cn">T</span>,
  seed <span class="op">=</span> <span class="fl">100</span>,
  path_folder <span class="op">=</span> <span class="st">'~/INDICA/xgb/cv2'</span>
<span class="op">)</span></code></pre></div>
<p>Note that many methods for processing data based on user-defined
parameters and machine learning-based methods are using functions from
the tidymodels (<a href="https://www.tidymodels.org/" class="uri">https://www.tidymodels.org/</a>) collection of R packages
(<span class="citation">Kuhn and Wickham (<a href="#ref-tidymodels" role="doc-biblioref">2020</a>)</span>).<br></p>
</div>
</div>
<div id="step-3-comes-soon" class="section level1">
<h1 class="hasAnchor">
<a href="#step-3-comes-soon" class="anchor"></a>Step 3: Comes soon!</h1>
</div>
<div id="step-4-comes-soon" class="section level1">
<h1 class="hasAnchor">
<a href="#step-4-comes-soon" class="anchor"></a>Step 4: Comes soon!</h1>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-chen2016xgboost" class="csl-entry">
Chen, Tianqi, and Carlos Guestrin. 2016. <span>“Xgboost: A Scalable Tree
Boosting System.”</span> In <em>Proceedings of the 22nd Acm Sigkdd
International Conference on Knowledge Discovery and Data Mining</em>,
785–94.
</div>
<div id="ref-costa2021nonlinear" class="csl-entry">
Costa-Neto, Germano, Roberto Fritsche-Neto, and José Crossa. 2021.
<span>“Nonlinear Kernels, Dominance, and Envirotyping Data Increase the
Accuracy of Genome-Based Prediction in Multi-Environment Trials.”</span>
<em>Heredity</em> 126 (1): 92–106.
</div>
<div id="ref-jarquin2014reaction" class="csl-entry">
Jarquı́n, Diego, José Crossa, Xavier Lacaze, Philippe Du Cheyron, Joëlle
Daucourt, Josiane Lorgeou, François Piraux, et al. 2014. <span>“A
Reaction Norm Model for Genomic Selection Using High-Dimensional Genomic
and Environmental Data.”</span> <em>Theoretical and Applied
Genetics</em> 127 (3): 595–607.
</div>
<div id="ref-jarquin2017increasing" class="csl-entry">
Jarquı́n, Diego, Cristiano Lemes da Silva, R Chris Gaynor, Jesse Poland,
Allan Fritz, Reka Howard, Sarah Battenfield, and Jose Crossa. 2017.
<span>“Increasing Genomic-Enabled Prediction Accuracy by Modeling
Genotype<span class="math inline">\(\times\)</span> Environment
Interactions in Kansas Wheat.”</span> <em>The Plant Genome</em> 10 (2):
1–15.
</div>
<div id="ref-tidymodels" class="csl-entry">
Kuhn, Max, and Hadley Wickham. 2020. <em>Tidymodels: A Collection of
Packages for Modeling and Machine Learning Using Tidyverse
Principles.</em> <a href="https://www.tidymodels.org">https://www.tidymodels.org</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
