<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Processing of a split object to get data ready to be used and fitted with
a DL_reg_2 (neural network) regression model. — DL_reg_2 • learnMET</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Processing of a split object to get data ready to be used and fitted with
a DL_reg_2 (neural network) regression model. — DL_reg_2" />
<meta property="og:description" content="The function processes a split object (training + test sets), according to
the configuration set by the user. For instance, genomic information is
incorporated according to the option set by the user.  A list of specific
environmental covariables to use can be provided.
A recipe is created using the package recipes, to specify additional
preprocessing steps, such as standardization based on the training set, with
same transformations used on the test set. Variables with null variance are
removed. If year effect is included, it is converted to dummy variables.
Further fitting on the training set with a deep learning model (see function
fit_cv_split.DL_reg_2()))." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_cv_stacking_indica.html">Evaluation of the model performance of a stacked model using CV0 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.html">Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_stacking.html">Overview on the different Machine Learning-based models in learnMET</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_2</code> (neural network) regression model.</h1>
    <small class="dont-index">Source: <a href='https://github.com/cjubin/learnMET/blob/master/R/DL_reg_2.R'><code>R/DL_reg_2.R</code></a></small>
    <div class="hidden name"><code>DL_reg_2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function processes a split object (training + test sets), according to
the configuration set by the user. For instance, genomic information is
incorporated according to the option set by the user.  A list of specific
environmental covariables to use can be provided.<br /></p>
<p>A recipe is created using the package <code>recipes</code>, to specify additional
preprocessing steps, such as standardization based on the training set, with
same transformations used on the test set. Variables with null variance are
removed. If year effect is included, it is converted to dummy variables.<br />
Further fitting on the training set with a deep learning model (see function
<code><a href='fit_cv_split.html'>fit_cv_split.DL_reg_2()</a></code>)).</p>
    </div>

    <pre class="usage"><span class='fu'>new_DL_reg_2</span><span class='op'>(</span>
  split <span class='op'>=</span> <span class='cn'>NULL</span>,
  trait <span class='op'>=</span> <span class='cn'>NULL</span>,
  geno <span class='op'>=</span> <span class='cn'>NULL</span>,
  env_predictors <span class='op'>=</span> <span class='cn'>NULL</span>,
  info_environments <span class='op'>=</span> <span class='cn'>NULL</span>,
  use_selected_markers <span class='op'>=</span> <span class='cn'>F</span>,
  SNPs <span class='op'>=</span> <span class='cn'>NULL</span>,
  include_env_predictors <span class='op'>=</span> <span class='cn'>T</span>,
  list_env_predictors <span class='op'>=</span> <span class='cn'>NULL</span>,
  lat_lon_included <span class='op'>=</span> <span class='cn'>F</span>,
  year_included <span class='op'>=</span> <span class='cn'>F</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='fu'>DL_reg_2</span><span class='op'>(</span>
  <span class='va'>split</span>,
  <span class='va'>trait</span>,
  <span class='va'>geno</span>,
  <span class='va'>env_predictors</span>,
  <span class='va'>info_environments</span>,
  <span class='va'>use_selected_markers</span>,
  <span class='va'>SNPs</span>,
  <span class='va'>list_env_predictors</span>,
  <span class='va'>include_env_predictors</span>,
  <span class='va'>lat_lon_included</span>,
  <span class='va'>year_included</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='fu'>validate_DL_reg_2</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>trait</th>
      <td><p><code>character</code> Name of the trait to predict. An ordinal trait
should be encoded as <code>integer</code>.</p></td>
    </tr>
    <tr>
      <th>geno</th>
      <td><p><code>data.frame</code> It corresponds to a <code>geno</code> element
within an object of class <code>METData</code>.</p></td>
    </tr>
    <tr>
      <th>env_predictors</th>
      <td><p><code>data.frame</code> It corresponds to the <code>env_data</code>
element within an object of class <code>METData</code>.</p></td>
    </tr>
    <tr>
      <th>info_environments</th>
      <td><p><code>data.frame</code> It corresponds to the
<code>info_environments</code> element within an object of class <code>METData</code>.</p></td>
    </tr>
    <tr>
      <th>use_selected_markers</th>
      <td><p>A <code>Logical</code> indicating whether to use a
subset of markers  identified via single-environment GWAS or based on the
table of marker effects obtained via Elastic Net as predictor variables,
when main genetic effects are modeled with principal components. <br />
If <code>use_selected_markers</code> is <code>TRUE</code>, the <code>SNPs</code> argument should be
provided.
<strong>For more details, see <code><a href='select_markers.html'>select_markers()</a></code></strong></p></td>
    </tr>
    <tr>
      <th>SNPs</th>
      <td><p>A <code>data.frame</code> with the genotype matrix (individuals in rows
and selected markers in columns) for SNPs selected via the
<code><a href='select_markers.html'>select_markers()</a></code> function.
<strong>Optional argument, can remain as <code>NULL</code> if no single markers should
be incorporated as predictor variables in analyses based on PCA
decomposition.</strong></p></td>
    </tr>
    <tr>
      <th>include_env_predictors</th>
      <td><p>A <code>logical</code> indicating whether
environmental covariates characterizing each environment should be used in
predictions.</p></td>
    </tr>
    <tr>
      <th>list_env_predictors</th>
      <td><p>A <code>character</code> vector containing the names
of the environmental predictors which should be used in predictions.
<strong>By default <code>NULL</code>: all environmental predictors included in the
env_data table of the <code>METData</code> object will be used.</strong></p></td>
    </tr>
    <tr>
      <th>lat_lon_included</th>
      <td><p><code>logical</code> indicates if longitude and latitude
data should be used as numeric predictors. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>year_included</th>
      <td><p><code>logical</code> indicates if year factor should be used
as predictor variable. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>an</th>
      <td><p>object of class <code>split</code>, which is a subelement of the output of the
<code><a href='predict_cv00.html'>predict_cv00()</a></code>, <code><a href='predict_cv0.html'>predict_cv0()</a></code>, <code><a href='predict_cv1.html'>predict_cv1()</a></code> and <code><a href='predict_cv2.html'>predict_cv2()</a></code>
functions. A <code>split</code> object contains a training and test elements.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>list</code> object of class <code>DL_reg_2</code> with the following items:</p><dl>
<dt>training</dt><dd><p><code>data.frame</code> Training set after partial processing</p></dd>
<dt>test</dt><dd><p><code>data.frame</code> Test set after partial processing</p></dd>
<dt>rec</dt><dd><p>A <code>recipe</code> object, specifying the remaining processing
steps which are implemented when a model is fitted on the training set
with a recipe.</p></dd>

</dl>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Wickham H, Averick M, Bryan J, Chang W, McGowan LD, Fran攼㸷ois R, Grolemund G, Hayes A, Henry L, Hester J, others (2019).
&#8220;Welcome to the Tidyverse.&#8221;
<em>Journal of open source software</em>, <b>4</b>(43), 1686.

Kuhn M, Wickham H (2020).
<em>Tidymodels: a collection of packages for modeling and machine learning using tidyverse principles.</em>.
<a href='https://www.tidymodels.org'>https://www.tidymodels.org</a>.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


