<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • learnMET</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_cv_stacking_indica.html">Evaluation of the model performance of a stacked model using CV0 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.html">Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_getweatherdata.html">Create a METData object with weather data retrieved by nasapower</a>
    </li>
    <li>
      <a href="../articles/vignette_stacking.html">Overview on the different Machine Learning-based models in learnMET</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-step-create-metdata-object" class="hasAnchor"><a href="#section-step-create-metdata-object" class="anchor"></a>Step 1: create METData object</h2>
          <p class="section-desc"><p>Specify input data and processing parameters (automatic retrieval of external weather data; QC on raw weather data, if provided…)</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-main-function" class="hasAnchor"><a href="#section-main-function" class="anchor"></a>Main function</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="create_METData.html">new_create_METData()</a></code> <code><a href="create_METData.html">create_METData()</a></code> <code><a href="create_METData.html">validate_create_METData()</a></code> </p>
        </td>
        <td><p>Create a multi-environment trials data object</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-get-daily-weather-data-for-an-environment-based-on-geographical-coordinates" class="hasAnchor"><a href="#section-get-daily-weather-data-for-an-environment-based-on-geographical-coordinates" class="anchor"></a>Get daily weather data for an environment based on geographical coordinates</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_daily_tables_per_env.html">get_daily_tables_per_env()</a></code> </p>
        </td>
        <td><p>Obtain daily climate data for an environment from NASA POWER data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="daylength.html">daylength()</a></code> </p>
        </td>
        <td><p>Compute the day length given the altitude and day of year.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.html">get.ea()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.with.rhmean.html">get.ea.with.rhmean()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.no.RH.html">get.ea.no.RH()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.es.html">get.es()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.esmn.html">get.esmn()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.esmx.html">get.esmx()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-check-daily-weather-data-non-exhaustive-quality-control-provided-by-user" class="hasAnchor"><a href="#section-check-daily-weather-data-non-exhaustive-quality-control-provided-by-user" class="anchor"></a>Check daily weather data (non-exhaustive quality control) provided by user</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="qc_raw_weather_data.html">qc_raw_weather_data()</a></code> </p>
        </td>
        <td><p>Quality control on daily weather data</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-compute-environmental-covariates-based-on-raw-daily-weather-data" class="hasAnchor"><a href="#section-compute-environmental-covariates-based-on-raw-daily-weather-data" class="anchor"></a>Compute environmental covariates based on raw daily weather data</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_ECs.html">get_ECs()</a></code> </p>
        </td>
        <td><p>Compute environmental covariates for each environment of the MET dataset.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_fixed_length_window.html">compute_EC_fixed_length_window()</a></code> </p>
        </td>
        <td><p>Compute ECs based on day-windows of fixed length.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_fixed_number_windows.html">compute_EC_fixed_number_windows()</a></code> </p>
        </td>
        <td><p>Compute ECs based on a fixed number of day-windows (fixed number across
all environments).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_gdd.html">compute_EC_gdd()</a></code> </p>
        </td>
        <td><p>Compute ECs based on growth stages which are estimated based on accumulated
GDD in each environment.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gdd_information.html">gdd_information()</a></code> </p>
        </td>
        <td><p>Internal function of [compute_EC_gdd())]

[compute_EC_gdd())]: R:compute_EC_gdd())</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_solar_radiation.html">get_solar_radiation()</a></code> </p>
        </td>
        <td><p>Obtain daily solar radiation for an IDenv with package nasapower
derived by NASA from satellite &amp; atmospheric observations.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_wind_data.html">get_wind_data()</a></code> </p>
        </td>
        <td><p>Obtain daily wind data for an IDenv with package nasapower
derived by NASA from satellite &amp; atmospheric observations.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-clustering-of-environments-based-on-weather-data-from-the-complete-training-dataset" class="hasAnchor"><a href="#section-clustering-of-environments-based-on-weather-data-from-the-complete-training-dataset" class="anchor"></a>Clustering of environments based on weather data from the complete training dataset</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="clustering_env_data.html">clustering_env_data()</a></code> </p>
        </td>
        <td><p>Clustering of environments solely based on environmental information</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-overview-of-the-metdata-object-created" class="hasAnchor"><a href="#section-overview-of-the-metdata-object-created" class="anchor"></a>Overview of the METData object created</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="summary.METData.html">summary(<i>&lt;METData&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summary of an object of class METData</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print.summary.METData.html">print(<i>&lt;summary.METData&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print the summary of an object of class METData</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-step-cross-validated-model-evaluation-of-the-metdata" class="hasAnchor"><a href="#section-step-cross-validated-model-evaluation-of-the-metdata" class="anchor"></a>Step 2: cross-validated model evaluation of the METData</h2>
          <p class="section-desc"><p>Evaluate predictive ability of a machine learning-based model with a specific CV scheme</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-main-function" class="hasAnchor"><a href="#section-main-function" class="anchor"></a>Main function</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="predict_trait_MET_cv.html">predict_trait_MET_cv()</a></code> </p>
        </td>
        <td><p>Cross-validation procedure for phenotypic prediction of crop varieties.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-create-train-test-splits-to-address-typical-prediction-problems-for-met-datasets" class="hasAnchor"><a href="#section-create-train-test-splits-to-address-typical-prediction-problems-for-met-datasets" class="anchor"></a>Create train/test splits to address typical prediction problems for MET datasets</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="predict_cv0.html">predict_cv0()</a></code> </p>
        </td>
        <td><p>Get train/test splits of the phenotypic MET dataset based on CV0.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="predict_cv00.html">predict_cv00()</a></code> </p>
        </td>
        <td><p>Get train/test splits of the phenotypic MET dataset based on CV0.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="predict_cv1.html">predict_cv1()</a></code> </p>
        </td>
        <td><p>Get train/test splits of the phenotypic MET dataset based on CV1.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="predict_cv2.html">predict_cv2()</a></code> </p>
        </td>
        <td><p>Get train/test splits of the phenotypic MET dataset based on CV2.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-processing-of-genotypic-data-for-ml-based-predictions" class="hasAnchor"><a href="#section-processing-of-genotypic-data-for-ml-based-predictions" class="anchor"></a>Processing of genotypic data for ML-based predictions</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="apply_pca.html">apply_pca()</a></code> </p>
        </td>
        <td><p>Data dimensionality reduction using PCA on a split object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_pcs_G_Add.html">apply_pcs_G_Add()</a></code> </p>
        </td>
        <td><p>Data dimensionality reduction by modeling genetic effects using the PCs of
the genomic relationship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_markers.html">select_markers()</a></code> </p>
        </td>
        <td><p>Selection of specific SNPs covariates.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_effect_per_env_EN.html">marker_effect_per_env_EN()</a></code> </p>
        </td>
        <td><p>Compute marker effects per environment with Elastic Net</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_effect_per_env_FarmCPU.html">marker_effect_per_env_FarmCPU()</a></code> </p>
        </td>
        <td><p>Compute marker P-values for each environment with GWAS.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-ml-methods-implemented-processing-functions-according-to-the-method" class="hasAnchor"><a href="#section-ml-methods-implemented-processing-functions-according-to-the-method" class="anchor"></a>ML-methods implemented: processing functions according to the method</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_splits_processed_with_method.html">get_splits_processed_with_method()</a></code> </p>
        </td>
        <td><p>Attribute a processing method for each list of training/test splits</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_1.html">new_stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">validate_stacking_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_1</code> (stacking of SVM models) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_2.html">new_stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">validate_stacking_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_3.html">new_stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">validate_stacking_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>rf_reg_1</code> (random forest) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_2.html">new_xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">validate_xgb_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_3.html">new_xgb_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_1.html">new_DL_reg_1()</a></code> <code><a href="DL_reg_1.html">DL_reg_1()</a></code> <code><a href="DL_reg_1.html">validate_DL_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_1</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_2.html">new_DL_reg_2()</a></code> <code><a href="DL_reg_2.html">DL_reg_2()</a></code> <code><a href="DL_reg_2.html">validate_DL_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_2</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_3.html">new_DL_reg_3()</a></code> <code><a href="DL_reg_3.html">DL_reg_3()</a></code> <code><a href="DL_reg_3.html">validate_DL_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_3</code> (neural network) regression model.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-ml-methods-implemented-fitting-functions-according-to-the-method" class="hasAnchor"><a href="#section-ml-methods-implemented-fitting-functions-according-to-the-method" class="anchor"></a>ML-methods implemented: fitting functions according to the method</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="fit_cv_split.html">fit_cv_split()</a></code> </p>
        </td>
        <td><p>S3 method used to fit an object of class <code>xgb_reg_1</code>, <code>xgb_reg_2</code>,
<code>xgb_reg_3</code>,<code>DL_reg</code>,<code>DL_reg_1</code>,<code>DL_reg_2</code>,<code>DL_reg_3</code>,<code>stacking_reg_1</code>,
<code>stacking_reg_2</code> or <code>stacking_reg_3</code>.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-compute-variable-importance-model-specific-and-model-free-e-g-permutation-based-methods-" class="hasAnchor"><a href="#section-compute-variable-importance-model-specific-and-model-free-e-g-permutation-based-methods-" class="anchor"></a>Compute variable importance (model-specific and model-free, e.g. permutation-based methods)</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="variable_importance_split.html">variable_importance_split()</a></code> </p>
        </td>
        <td><p>Compute variable importance according to the machine learning algorithm used</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-plot-cross-validated-results-for-predictive-ability" class="hasAnchor"><a href="#section-plot-cross-validated-results-for-predictive-ability" class="anchor"></a>Plot cross-validated results for predictive ability</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_results_cv.html">plot_results_cv()</a></code> </p>
        </td>
        <td><p>bdyy</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-plot-variable-importance-results" class="hasAnchor"><a href="#section-plot-variable-importance-results" class="anchor"></a>Plot variable importance results</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_results_vip_cv.html">plot_results_vip_cv()</a></code> </p>
        </td>
        <td><p>Plot variable importance scores</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-step-create-a-table-of-new-phenotypes-to-predict-i-e-which-genotypes-in-which-environments" class="hasAnchor"><a href="#section-step-create-a-table-of-new-phenotypes-to-predict-i-e-which-genotypes-in-which-environments" class="anchor"></a>Step 3: Create a table of new phenotypes to predict (i.e. which genotypes in which environments</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-main-function" class="hasAnchor"><a href="#section-main-function" class="anchor"></a>Main function</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="add_METData_to_predict.html">new_add_METData_to_predict()</a></code> <code><a href="add_METData_to_predict.html">add_METData_to_predict()</a></code> <code><a href="add_METData_to_predict.html">validate_add_METData_to_predict()</a></code> </p>
        </td>
        <td><p>Create a multi-environment trials data object, which will be used for
prediction of unobserved data.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-get-daily-weather-data-for-an-environment-based-on-geographical-coordinates" class="hasAnchor"><a href="#section-get-daily-weather-data-for-an-environment-based-on-geographical-coordinates" class="anchor"></a>Get daily weather data for an environment based on geographical coordinates</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_daily_tables_per_env.html">get_daily_tables_per_env()</a></code> </p>
        </td>
        <td><p>Obtain daily climate data for an environment from NASA POWER data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="daylength.html">daylength()</a></code> </p>
        </td>
        <td><p>Compute the day length given the altitude and day of year.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.html">get.ea()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.with.rhmean.html">get.ea.with.rhmean()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ea.no.RH.html">get.ea.no.RH()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.es.html">get.es()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.esmn.html">get.esmn()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.esmx.html">get.esmx()</a></code> </p>
        </td>
        <td><p>Formulas to compute vapour pressure deficit according to available data</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-check-daily-weather-data-non-exhaustive-quality-control-provided-by-user" class="hasAnchor"><a href="#section-check-daily-weather-data-non-exhaustive-quality-control-provided-by-user" class="anchor"></a>Check daily weather data (non-exhaustive quality control) provided by user</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="qc_raw_weather_data.html">qc_raw_weather_data()</a></code> </p>
        </td>
        <td><p>Quality control on daily weather data</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-compute-environmental-covariates-based-on-raw-daily-weather-data" class="hasAnchor"><a href="#section-compute-environmental-covariates-based-on-raw-daily-weather-data" class="anchor"></a>Compute environmental covariates based on raw daily weather data</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_ECs.html">get_ECs()</a></code> </p>
        </td>
        <td><p>Compute environmental covariates for each environment of the MET dataset.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_fixed_length_window.html">compute_EC_fixed_length_window()</a></code> </p>
        </td>
        <td><p>Compute ECs based on day-windows of fixed length.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_fixed_number_windows.html">compute_EC_fixed_number_windows()</a></code> </p>
        </td>
        <td><p>Compute ECs based on a fixed number of day-windows (fixed number across
all environments).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_EC_gdd.html">compute_EC_gdd()</a></code> </p>
        </td>
        <td><p>Compute ECs based on growth stages which are estimated based on accumulated
GDD in each environment.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gdd_information.html">gdd_information()</a></code> </p>
        </td>
        <td><p>Internal function of [compute_EC_gdd())]

[compute_EC_gdd())]: R:compute_EC_gdd())</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_solar_radiation.html">get_solar_radiation()</a></code> </p>
        </td>
        <td><p>Obtain daily solar radiation for an IDenv with package nasapower
derived by NASA from satellite &amp; atmospheric observations.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_wind_data.html">get_wind_data()</a></code> </p>
        </td>
        <td><p>Obtain daily wind data for an IDenv with package nasapower
derived by NASA from satellite &amp; atmospheric observations.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-clustering-of-environments-based-on-weather-data-from-the-complete-training-dataset" class="hasAnchor"><a href="#section-clustering-of-environments-based-on-weather-data-from-the-complete-training-dataset" class="anchor"></a>Clustering of environments based on weather data from the complete training dataset</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="clustering_env_data.html">clustering_env_data()</a></code> </p>
        </td>
        <td><p>Clustering of environments solely based on environmental information</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-step-prediction-of-performance-for-untested-genotypes-and-or-environment" class="hasAnchor"><a href="#section-step-prediction-of-performance-for-untested-genotypes-and-or-environment" class="anchor"></a>Step 4: Prediction of performance for untested genotypes and/or environment</h2>
          <p class="section-desc"><p>Implement predictions for unobserved configurations of genotypic and environmental predictors</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-main-function" class="hasAnchor"><a href="#section-main-function" class="anchor"></a>Main function</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="predict_trait_MET.html">predict_trait_MET()</a></code> </p>
        </td>
        <td><p>Phenotypic prediction of unobserved data.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-processing-of-genotypic-data-for-ml-based-predictions" class="hasAnchor"><a href="#section-processing-of-genotypic-data-for-ml-based-predictions" class="anchor"></a>Processing of genotypic data for ML-based predictions</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="apply_pca.html">apply_pca()</a></code> </p>
        </td>
        <td><p>Data dimensionality reduction using PCA on a split object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_pcs_G_Add.html">apply_pcs_G_Add()</a></code> </p>
        </td>
        <td><p>Data dimensionality reduction by modeling genetic effects using the PCs of
the genomic relationship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_markers.html">select_markers()</a></code> </p>
        </td>
        <td><p>Selection of specific SNPs covariates.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_effect_per_env_EN.html">marker_effect_per_env_EN()</a></code> </p>
        </td>
        <td><p>Compute marker effects per environment with Elastic Net</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="marker_effect_per_env_FarmCPU.html">marker_effect_per_env_FarmCPU()</a></code> </p>
        </td>
        <td><p>Compute marker P-values for each environment with GWAS.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-ml-methods-implemented-processing-functions-according-to-the-method" class="hasAnchor"><a href="#section-ml-methods-implemented-processing-functions-according-to-the-method" class="anchor"></a>ML-methods implemented: processing functions according to the method</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_splits_processed_with_method.html">get_splits_processed_with_method()</a></code> </p>
        </td>
        <td><p>Attribute a processing method for each list of training/test splits</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_1.html">new_stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">validate_stacking_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_1</code> (stacking of SVM models) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_2.html">new_stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">validate_stacking_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_3.html">new_stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">validate_stacking_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>rf_reg_1</code> (random forest) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_2.html">new_xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">validate_xgb_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_3.html">new_xgb_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_1.html">new_DL_reg_1()</a></code> <code><a href="DL_reg_1.html">DL_reg_1()</a></code> <code><a href="DL_reg_1.html">validate_DL_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_1</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_2.html">new_DL_reg_2()</a></code> <code><a href="DL_reg_2.html">DL_reg_2()</a></code> <code><a href="DL_reg_2.html">validate_DL_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_2</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_3.html">new_DL_reg_3()</a></code> <code><a href="DL_reg_3.html">DL_reg_3()</a></code> <code><a href="DL_reg_3.html">validate_DL_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_3</code> (neural network) regression model.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-ml-methods-implemented-fitting-functions-according-to-the-method" class="hasAnchor"><a href="#section-ml-methods-implemented-fitting-functions-according-to-the-method" class="anchor"></a>ML-methods implemented: fitting functions according to the method</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="fit_predict_model.html">fit_predict_model()</a></code> </p>
        </td>
        <td><p>fit_predict_model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_1.html">new_stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">stacking_reg_1()</a></code> <code><a href="stacking_reg_1.html">validate_stacking_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_1</code> (stacking of SVM models) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_2.html">new_stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">stacking_reg_2()</a></code> <code><a href="stacking_reg_2.html">validate_stacking_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stacking_reg_3.html">new_stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">stacking_reg_3()</a></code> <code><a href="stacking_reg_3.html">validate_stacking_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>stacking_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">new_xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">xgb_reg_1()</a></code> <code><a href="xgb_reg_1.html">validate_xgb_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>rf_reg_1</code> (random forest) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_2.html">new_xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">xgb_reg_2()</a></code> <code><a href="xgb_reg_2.html">validate_xgb_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_2</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xgb_reg_3.html">new_xgb_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>xgb_reg_3</code> (gradient boosted tree) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_1.html">new_DL_reg_1()</a></code> <code><a href="DL_reg_1.html">DL_reg_1()</a></code> <code><a href="DL_reg_1.html">validate_DL_reg_1()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_1</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_2.html">new_DL_reg_2()</a></code> <code><a href="DL_reg_2.html">DL_reg_2()</a></code> <code><a href="DL_reg_2.html">validate_DL_reg_2()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_2</code> (neural network) regression model.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DL_reg_3.html">new_DL_reg_3()</a></code> <code><a href="DL_reg_3.html">DL_reg_3()</a></code> <code><a href="DL_reg_3.html">validate_DL_reg_3()</a></code> </p>
        </td>
        <td><p>Processing of a split object to get data ready to be used and fitted with
a <code>DL_reg_3</code> (neural network) regression model.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-compute-variable-importance-model-specific-and-model-free-e-g-permutation-based-methods-" class="hasAnchor"><a href="#section-compute-variable-importance-model-specific-and-model-free-e-g-permutation-based-methods-" class="anchor"></a>Compute variable importance (model-specific and model-free, e.g. permutation-based methods)</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="variable_importance_split.html">variable_importance_split()</a></code> </p>
        </td>
        <td><p>Compute variable importance according to the machine learning algorithm used</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-plot-variable-importance-results" class="hasAnchor"><a href="#section-plot-variable-importance-results" class="anchor"></a>Plot variable importance results</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_results_vip.html">plot_results_vip()</a></code> </p>
        </td>
        <td><p>Plot variable importance scores</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-step-analysis-of-prediction-results-for-new-observations-by-location-and-by-environmental-cluster" class="hasAnchor"><a href="#section-step-analysis-of-prediction-results-for-new-observations-by-location-and-by-environmental-cluster" class="anchor"></a>Step 5: Analysis of prediction results for new observations, by location and by environmental cluster</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="analysis_predictions_best_genotypes.html">analysis_predictions_best_genotypes()</a></code> </p>
        </td>
        <td><p>Analysis of prediction results from <code>predict_trait_MET()</code> by location and by
environmental cluster.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-datasets" class="hasAnchor"><a href="#section-datasets" class="anchor"></a>Datasets</h2>
          <p class="section-desc"><p>Toy data to illustrate the use of the package functions</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pheno_indica.html">pheno_indica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geno_indica.html">geno_indica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_indica.html">map_indica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="climate_variables_indica.html">climate_variables_indica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="info_environments_indica.html">info_environments_indica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pheno_japonica.html">pheno_japonica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geno_japonica.html">geno_japonica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_japonica.html">map_japonica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="climate_variables_japonica.html">climate_variables_japonica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="info_environments_japonica.html">info_environments_japonica</a></code> </p>
        </td>
        <td><p>Multi-year trial data of rice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pheno_G2F.html">pheno_G2F</a></code> </p>
        </td>
        <td><p>Maize experimental multi-environment data sets (Genomes to Fields Initiative)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geno_G2F.html">geno_G2F</a></code> </p>
        </td>
        <td><p>Maize experimental multi-environment data sets (Genomes to Fields Initiative)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="map_G2F.html">map_G2F</a></code> </p>
        </td>
        <td><p>Maize experimental multi-environment data sets (Genomes to Fields Initiative)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="soil_G2F.html">soil_G2F</a></code> </p>
        </td>
        <td><p>Maize experimental multi-environment data sets (Genomes to Fields Initiative)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="info_environments_G2F.html">info_environments_G2F</a></code> </p>
        </td>
        <td><p>Maize experimental multi-environment data sets (Genomes to Fields Initiative)</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


