<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute marker P-values for each environment with GWAS. — marker_effect_per_env_FarmCPU • learnMET</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Compute marker P-values for each environment with GWAS. — marker_effect_per_env_FarmCPU" />
<meta property="og:description" content="GWAS method implemented: FarmCPU from GAPIT3.
Multiple testing correction: Benjamini–Hochberg procedure with alpha=0.05." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_stacking.html">Overview on the different Machine Learning-based methods</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_stacking_indica.Rmd">Evaluation of the performance of a stacked model using CV0 for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.Rmd">Evaluation of the performance of a XGBoost model using CV2 for phenotypic prediction</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute marker P-values for each environment with GWAS.</h1>
    <small class="dont-index">Source: <a href='https://github.com/cjubin/learnMET/blob/master/R/marker_effect_per_env_FarmCPU.R'><code>R/marker_effect_per_env_FarmCPU.R</code></a></small>
    <div class="hidden name"><code>marker_effect_per_env_FarmCPU.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>GWAS method implemented: FarmCPU from GAPIT3.
Multiple testing correction: Benjamini–Hochberg procedure with alpha=0.05.</p>
    </div>

    <pre class="usage"><span class='fu'>marker_effect_per_env_FarmCPU</span><span class='op'>(</span>
  <span class='va'>geno</span>,
  <span class='va'>pheno</span>,
  <span class='va'>map</span>,
  <span class='va'>environment</span>,
  <span class='va'>pheno_trait</span>,
  nb_pcs <span class='op'>=</span> <span class='fl'>5</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>geno</th>
      <td><p><code>data.frame</code>with markers in columns and inviduals in rows.
Typical input is <code>METData$geno</code> from the <code><a href='create_METData.html'>create_METData()</a></code> function.</p></td>
    </tr>
    <tr>
      <th>pheno</th>
      <td><p><code>data.frame</code> object with at least 4 columns.</p><ol>
<li><p>geno_ID <code>character</code> contains the genotype identifiers.</p></li>
<li><p>year <code>numeric</code> contains the year of the observation.</p></li>
<li><p>location <code>character</code> contains the name of the location.</p></li>
</ol>
<ul>
<li><p><strong>From the fourth column on: each column <code>numeric</code> contains
phenotypic values for a phenotypic trait observed in a combination
Year x Location. Names of the traits can be provided as column names.</strong></p></li>
<li><p><strong>Last column: IDenv (combination LocationxYear)</strong></p></li>
</ul>

<p>Typical input is <code>METData$pheno</code> from the <code><a href='create_METData.html'>create_METData()</a></code> function.</p></td>
    </tr>
    <tr>
      <th>map</th>
      <td><p><code>data.frame</code> with 3 columns.</p><ol>
<li><p>marker_name <code>character</code> with marker names</p></li>
<li><p>chr <code>numeric</code> with chromosome number</p></li>
<li><p>pos <code>numeric</code> with marker position.</p></li>
</ol><p>Typical input is METData$map from the <code><a href='create_METData.html'>create_METData()</a></code> function.</p></td>
    </tr>
    <tr>
      <th>environment</th>
      <td><p><code>character</code> indicating the name of the environment
for which marker effects should be computed</p></td>
    </tr>
    <tr>
      <th>pheno_trait</th>
      <td><p><code>character</code> Name of the trait under study for which marker
effects should be estimated</p></td>
    </tr>
    <tr>
      <th>nb_pcs</th>
      <td><p><code>numeric</code> Number of principal components to use in FarmCPU
analyses.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>list</code> which contains the following elements:</p><dl>
<dt>GWAS_results</dt><dd><p><code>data.frame</code> FarmCPU results for all SNPs</p></dd>
<dt>threshold</dt><dd><p><code>numeric</code> Cutoff derived from Benjamini-Hochberg
procedure</p></dd>
<dt>selected_markers</dt><dd><p><code>character</code> Vector containing the names of
the SNPs passing the threshold</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For more information about the GWAS method, please consult the original
publication from the authors:
Liu, Xiaolei, et al. "Iterative usage of fixed and random effect models for
powerful and efficient genome-wide association studies."
PLoS genetics 12.2 (2016): e1005767.
<a href='https://doi.org/10.1371/journal.pgen.1005767'>https://doi.org/10.1371/journal.pgen.1005767</a></p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Cathy C. Westhues <a href='mailto:cathy.jubin@uni-goettingen.de'>cathy.jubin@uni-goettingen.de</a></p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


