<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Phenotypic prediction of unobserved data. — predict_trait_MET • learnMET</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Phenotypic prediction of unobserved data. — predict_trait_MET" />
<meta property="og:description" content="Implement trait prediction based on SNP and environmental data
with selection of prediction methods among Machine Learning approaches.
This function should be used to assess the predictive ability according to
a cross-validation scheme determined by the user." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">learnMET</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_cv_stacking_indica.html">Evaluation of the model performance of a stacked model using CV0 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_cv_xgb_indica.html">Evaluation of the performance of a XGBoost model using CV2 on a rice dataset for phenotypic prediction</a>
    </li>
    <li>
      <a href="../articles/vignette_getweatherdata.html">Create a METData object with weather data retrieved by nasapower</a>
    </li>
    <li>
      <a href="../articles/vignette_stacking.html">Overview on the different Machine Learning-based models in learnMET</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cjubin/learnMET/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Phenotypic prediction of unobserved data.</h1>
    <small class="dont-index">Source: <a href='https://github.com/cjubin/learnMET/blob/master/R/predict_trait_MET.R'><code>R/predict_trait_MET.R</code></a></small>
    <div class="hidden name"><code>predict_trait_MET.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Implement trait prediction based on SNP and environmental data
with selection of prediction methods among Machine Learning approaches.</p>
<p>This function should be used to assess the predictive ability according to
a cross-validation scheme determined by the user.</p>
    </div>

    <pre class="usage"><span class='fu'>predict_trait_MET</span><span class='op'>(</span>
  <span class='va'>METData_training</span>,
  <span class='va'>METData_new</span>,
  <span class='va'>trait</span>,
  prediction_method <span class='op'>=</span> <span class='st'>"xgb_reg"</span>,
  use_selected_markers <span class='op'>=</span> <span class='cn'>F</span>,
  list_selected_markers_manual <span class='op'>=</span> <span class='cn'>NULL</span>,
  geno_information <span class='op'>=</span> <span class='st'>"PCs"</span>,
  lat_lon_included <span class='op'>=</span> <span class='cn'>F</span>,
  year_included <span class='op'>=</span> <span class='cn'>F</span>,
  include_env_predictors <span class='op'>=</span> <span class='cn'>T</span>,
  list_env_predictors <span class='op'>=</span> <span class='cn'>NULL</span>,
  seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  save_processing <span class='op'>=</span> <span class='cn'>T</span>,
  <span class='va'>path_folder</span>,
  compute_vip <span class='op'>=</span> <span class='cn'>TRUE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>trait</th>
      <td><p><code>character</code> Name of the trait to predict. An ordinal trait
should be encoded as <code>integer</code>.</p></td>
    </tr>
    <tr>
      <th>prediction_method</th>
      <td><p><code>character</code> specifying the predictive model to use.
Options are <code>xgb_reg</code> (gradient boosted trees),
(stacking of support vector machines with LASSO as meta-learner).</p></td>
    </tr>
    <tr>
      <th>use_selected_markers</th>
      <td><p>A <code>Logical</code> indicating whether to use a
subset of markers obtained from a previous step
(see <code>function select_markers()</code>).</p></td>
    </tr>
    <tr>
      <th>geno_information</th>
      <td><p>indicating how the complete genotypic matrix should
be used in predictions. Options are <code>SNPs</code> (all
of the markers will be individually used), <code>PCs</code> (PCA will be applied on
each genotype matrix for the training set for dimensionality reduction)
or <code>PCs_G</code> (decomposition of the genomic relationship matrix via PCA -- not
yet implemented).</p></td>
    </tr>
    <tr>
      <th>lat_lon_included</th>
      <td><p><code>logical</code> indicates if longitude and latitude
data should be used as numeric predictors. Default is <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>year_included</th>
      <td><p><code>logical</code> indicates if year factor should be used
as predictor variable. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>include_env_predictors</th>
      <td><p>A <code>logical</code> indicating whether
environmental covariates characterizing each environment should be used in
predictions.</p></td>
    </tr>
    <tr>
      <th>list_env_predictors</th>
      <td><p>A <code>character</code> vector containing the names
of the environmental predictors which should be used in predictions. By
default <code>NULL</code>: all environmental predictors included in the env_data table
of the <code>METData</code> object will be used.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p><code>integer</code> Seed value. Default is <code>NULL</code>. By default, a
random seed will be generated.</p></td>
    </tr>
    <tr>
      <th>save_processing</th>
      <td><p>a <code>logical</code> indicating whether the processing
steps obtained from the <code>processing_train_test_split()</code> or
<code>processing_train_test_split_kernel()</code> functions should be saved in a .RDS
object. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>path_folder</th>
      <td><p>a <code>character</code> indicating the full path where the .RDS
object and plots generated during the analysis should be saved (do not use
a Slash after the name of the last folder). Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed to the <code>processing_train_test_split()</code>,
<code>processing_train_test_split_kernel()</code>, <code>reg_fitting_train_test_split()</code>,
<code>reg_fitting_train_test_split_kernel()</code> functions.</p></td>
    </tr>
    <tr>
      <th>METData</th>
      <td><p><code>list</code> An object created by the initial function of the
package create_METData().</p></td>
    </tr>
    <tr>
      <th>cv_type</th>
      <td><p>A <code>character</code> with one out of <code>cv0</code> (prediction of new
environments), <code>cv00</code> (prediction of new genotypes in new environments),
<code>cv1</code> (prediction of new genotypes) or <code>cv2</code> (prediction of incomplete
field trials). Default is <code>cv0</code>.</p></td>
    </tr>
    <tr>
      <th>cv0_type</th>
      <td><p>A <code>character</code> with one out of
<code>leave-one-environment-out</code>, <code>leave-one-site-out</code>,<code>leave-one-year-out</code>,
<code>forward-prediction</code>. Default is <code>leave-one-environment-out</code>.</p></td>
    </tr>
    <tr>
      <th>nb_folds_cv1</th>
      <td><p>A <code>numeric</code> Number of folds used in the CV1 scheme.
Default is 5.</p></td>
    </tr>
    <tr>
      <th>repeats_cv1</th>
      <td><p>A <code>numeric</code> Number of repeats in the CV1 scheme.
Default is 50.</p></td>
    </tr>
    <tr>
      <th>nb_folds_cv2</th>
      <td><p>A <code>numeric</code> Number of folds used in the CV2 scheme.
Default is 5.</p></td>
    </tr>
    <tr>
      <th>repeats_cv2</th>
      <td><p>A <code>numeric</code> Number of repeats in the CV2 scheme.
Default is 50.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>list</code> object of class <code>met_cv</code> with the following items:</p><dl>
<dt>list_results_cv</dt><dd><p><code>list</code> of <code>res_fitted_split</code> elements.
Detailed prediction results for each split of the
data within each element of this list.</p></dd>
<dt>seed_used</dt><dd><p><code>integer</code> Seed used to generate the
cross-validation splits.</p></dd>

</dl>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Cathy C. Westhues <a href='mailto:cathy.jubin@uni-goettingen.de'>cathy.jubin@uni-goettingen.de</a></p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Cathy Westhues (Jubin).</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


